server/src/main/java/org/elasticsearch/snapshots/RestoreService.java

```diff
@@ -235,492 +234,152 @@ public void restoreSnapshot(final RestoreSnapshotRequest request,
-                                    if (RepositoriesMetadata.TYPE.equals(cursor.key) == false
-                                            && DataStreamMetadata.TYPE.equals(cursor.key) == false
-                                            && cursor.value instanceof Metadata.NonRestorableCustom == false) {
-                                        // TODO: Check request.skipOperatorOnly for Autoscaling policies (NonRestorableCustom)
-                                        // Don't restore repositories while we are working with them
-                                        // TODO: Should we restore them at the end?
-                                        // Also, don't restore data streams here, we already added them to the metadata builder above
-                                        mdBuilder.putCustom(cursor.key, cursor.value);
+            repositoryDataListener.whenComplete(repositoryData ->
+                repositoryUuidRefreshListener.whenComplete(ignored ->
+                    // fork handling to the generic pool since it loads various pieces of metadata from the repository over a longer period
+                    // of time
+                    clusterService.getClusterApplierService().threadPool().generic().execute(
+                            ActionRunnable.wrap(
+                                listener,
+                                l -> {
+                                    final String snapshotName = request.snapshot();
+                                    final Optional<SnapshotId> matchingSnapshotId = repositoryData.getSnapshotIds().stream()
+                                            .filter(s -> snapshotName.equals(s.getName())).findFirst();
+                                    if (matchingSnapshotId.isPresent() == false) {
+                                        throw new SnapshotRestoreException(repositoryName, snapshotName, "snapshot does not exist");
                                     }
-                                }
-                            }
-                        }
-
-                        if (completed(shards)) {
-                            // We don't have any indices to restore - we are done
-                            restoreInfo = new RestoreInfo(snapshotId.getName(),
-                                List.copyOf(indices.keySet()),
-                                shards.size(),
-                                shards.size() - failedShards(shards));
-                        }
-
-                        RoutingTable rt = rtBuilder.build();
-                        updater.accept(currentState, mdBuilder);
-                        ClusterState updatedState = builder.metadata(mdBuilder).blocks(blocks).routingTable(rt).build();
-                        return allocationService.reroute(updatedState, "restored snapshot [" + snapshot + "]");
-                    }
-
-                    private void checkAliasNameConflicts(Map<String, String> renamedIndices, Set<String> aliases) {
```

Comment:
The way this works was adjusted here. No need to collect all aliases into a set up front and then check against the `renamedIndices` keyset when `renamedIndices` is constant throughout the process. Just check alias names against the keyset in-line as we loop through the indices.

Reply:
Could you explain why we check this one case here? There are other possible problems with aliases too: the new aliases might conflict with the names of existing indices and/or the names of the indices being restored might conflict with existing aliases. I mean we do check these things in `Metadata.Builder#build()` I’m just wondering why we check this one extra case explicitly like this too.